lib/spack/spack/cmd/container.py

```diff
@@ -0,0 +1,151 @@
+except ImportError:
+    tty.die("spack container requires HPCCM")
+
+import spack
+import spack.cmd
+import spack.spec
+
+description = "generate a container specification file"
+section = "build"
+level = "short"
+
+
+def setup_parser(subparser):
+    subparser.add_argument(
+        'specs', nargs=argparse.REMAINDER, help="specs of packages")
+
+    subparser.add_argument(
+        '--baseimage', action='store', type=str, default='centos:7',
+        help='container base image (default: centos:7)')
+    subparser.add_argument(
+        '--depth', action='store', type=int, default=2,
+        help='spack DAG depth to use (default: 2)')
+    subparser.add_argument(
+        '--distro', action='store', type=str, default='',
+        choices=['centos', 'rhel', 'ubuntu', 'ubuntu16', 'ubuntu18'],
+        help='Linux distribution type of the base image '
+        '(default: try to determine automatically from the base image name)')
+    subparser.add_argument(
+        '--format', action='store', type=str, default='docker',
+        choices=['docker', 'singularity'],
+        help='container specification format (default: docker)')
+    subparser.add_argument(
+        '--multi-stage', action='store_true', default=False,
+        help='generate a multi-stage Dockerfile (Docker specific)')
+    subparser.add_argument(
+        '--no-checksum', action='store_true', default=False,
+        help='do not use checksums to verify downloaded files (unsafe)')
+    subparser.add_argument(
+        '--show-log-on-error', action='store_true', default=False,
+        help='print full build log if build fails')
+    subparser.add_argument(
+        '--spack-branch', '--branch', action='store', type=str,
+        default='master',
+        help='branch of Spack to deploy (default: master)')
+
+
+def container(parser, args):
+    if not args.specs:
+        tty.die("spack spec requires at least one spec")
+
+    recipe = hpccm.Stage()
+
+    recipe += baseimage(image=args.baseimage, _distro=args.distro)
+
+    # OS dependencies
+    recipe += hpccm.building_blocks.packages(
+        apt=['autoconf', 'build-essential', 'bzip2', 'ca-certificates',
+             'coreutils', 'cpio', 'curl', 'environment-modules', 'git',
+             'gzip', 'libssl-dev', 'make', 'openssh-client', 'patch',
+             'pkg-config', 'svn', 'tar', 'tcl', 'unzip', 'zlib1g'],
+        yum=['autoconf', 'bzip2', 'ca-certificates', 'coreutils', 'cpio',
+             'curl', 'environment-modules', 'git', 'gzip', 'make',
+             'openssh-clients', 'openssl-devel', 'patch', 'pkg-config',
+             'svn', 'tar', 'tcl', 'unzip', 'zlib-devel'])
+
+    # GNU compilers
+    recipe += hpccm.building_blocks.gnu()
+
+    # Spack itself
+    spack_bb = bb_base()
+    spack_bb += comment('Spack')
+    spack_bb += shell(commands=[
+        'mkdir -p /opt && cd /opt',
+        'git clone --depth=1 --branch {} https://github.com/spack/spack'.format(args.spack_branch),
```

Comment:
Using a local copy Spack is a reasonable use case, especially for the package definition case.  

However, I'm not sure how to best handle this.  

Copying the local Spack copy into the container would copy all the locally installed packages if they are also installed in the Spack directory, which would be undesirable, especially if the container image is significantly different from the host (e.g., Ubuntu container image and an CentOS host).

Another option would be to a command line option to specify a file to copy from the host into the container image?  `spack container --copy-file path/on/host/file:/path/in/container/file`

If the user's work is checked into a branch, they could use `--spack-branch`.  This might be fine for some cases?

Reply:
BTW: the `{}`-style format strings have been replaced with old style format strings.

Reply:
> If the user's work is checked into a branch, they could use --spack-branch. This might be fine for some cases?

That's fine for this PR.

You may be able to use git clone on the local repo, which would not include files that aren't part of the SCM (including installed packages).